### ISSUE TITLE
"Unexpected token" error when failing to resolve constructor in irrefutable patterns

### ISSUE THREAD

**Issue by SystemFw** [unisonweb/unison#3982]
Consider these two snippets that use irrefutable patterns.

```haskell
foo myThing = 
  (MyThing thing) =  myThing
   ...
```
```haskell
foo myThing =
  t @ (MyThing thing) = myThing
  ...
```

If ucm cannot resolve `MyThing` for whatever reason (e.g. type-directed name resolution fails),  you get an `unexpected =` in the first code snippet, and an `unexpected @` in the second code snippet, which misleads you into thinking there is a syntax error.

**Comment by timjs** [comment]
Yeah, rubbed into the same issue. Actually thought this kind of pattern matching is not allowed in Unison but good to read it actually is!
